File: bindings/python/smoke_primitives.py
from __future__ import annotations

import json

import pqcb


kem_public_key, kem_secret_key = pqcb.kem_keypair()
ciphertext, encapsulated = pqcb.kem_encapsulate(kem_public_key)
decapsulated = pqcb.kem_decapsulate(kem_secret_key, ciphertext)

if encapsulated != decapsulated:
    raise RuntimeError("ML-KEM shared secrets did not match")

signature_public_key, signature_secret_key = pqcb.signature_keypair()
message = b"pqcb python primitive smoke"
signature = pqcb.sign(signature_secret_key, message)
pqcb.verify(signature_public_key, message, signature)

try:
    pqcb.verify(signature_public_key, b"tampered", signature)
    raise RuntimeError("tampered ML-DSA signature unexpectedly verified")
except pqcb.VerificationFailedError:
    pass

print(
    json.dumps(
        {
            "kemSharedSecretLength": len(encapsulated),
            "signatureLength": len(signature),
            "tamperedSignatureRejected": True,
        },
        sort_keys=True,
    )
)

File: bindings/node/src/smoke-primitives.ts
import { kem, signature, VerificationFailedError } from "./index.js";

console.error("node primitive smoke: kem keypair");
const kemKeypair = kem.keypair();
console.error("node primitive smoke: kem encapsulate");
const encapsulation = kem.encapsulate(kemKeypair.publicKey);
console.error("node primitive smoke: kem decapsulate");
const decapsulated = kem.decapsulate(
  kemKeypair.secretKey,
  encapsulation.ciphertext,
);

if (!encapsulation.sharedSecret.equals(decapsulated)) {
  throw new Error("ML-KEM shared secrets did not match");
}

console.error("node primitive smoke: signature keypair");
const signatureKeypair = signature.keypair();
const message = Buffer.from("pqcb node primitive smoke", "utf8");
console.error("node primitive smoke: signature sign");
const signatureBytes = signature.sign(signatureKeypair.secretKey, message);

console.error("node primitive smoke: signature verify");
signature.verify(signatureKeypair.publicKey, message, signatureBytes);

try {
  console.error("node primitive smoke: signature tamper check");
  signature.verify(
    signatureKeypair.publicKey,
    Buffer.from("tampered", "utf8"),
    signatureBytes,
  );
  throw new Error("tampered ML-DSA signature unexpectedly verified");
} catch (error) {
  if (!(error instanceof VerificationFailedError)) {
    throw error;
  }
}

console.log(
  JSON.stringify({
    kemSharedSecretLength: encapsulation.sharedSecret.length,
    signatureLength: signatureBytes.length,
    tamperedSignatureRejected: true,
  }),
);
